cores/esp8266/Esp.cpp

```diff
@@ -569,9 +578,58 @@ bool EspClass::flashEraseSector(uint32_t sector) {
     return rc == 0;
 }
 
+#if PUYA_SUPPORT
+static int spi_flash_write_puya(uint32_t offset, uint32_t *data, size_t size) {
+    // PUYA flash chips need to read existing data, update in memory and write modified data again.
+    static uint32_t *flash_write_puya_buf = 0;
+    int rc = 0;
+    uint32_t* ptr = data;
+
+    if (flash_write_puya_buf == 0) {
+        flash_write_puya_buf = (uint32_t*) malloc(PUYA_BUFFER_SIZE);
+        // No need to ever free this, since the flash chip will never change at runtime.
+        if (flash_write_puya_buf == 0) {
+            // Memory could not be allocated.
+            return 1; // SPI_FLASH_RESULT_ERR
+        }
+    }
+    size_t bytesLeft = size;
+    uint32_t pos = offset;
+    while (bytesLeft > 0 && rc == 0) {
+        size_t bytesNow = bytesLeft;
+        if (bytesNow > PUYA_BUFFER_SIZE) {
+            bytesNow = PUYA_BUFFER_SIZE;
+            bytesLeft -= PUYA_BUFFER_SIZE;
+        } else {
+            bytesLeft = 0;
+        }
+        rc = spi_flash_read(pos, flash_write_puya_buf, bytesNow);
+        if (rc != 0) {
+            return rc;
+        }
+        for (size_t i = 0; i < bytesNow / 4; ++i) {
+            flash_write_puya_buf[i] &= *ptr;
+            ++ptr;
+        }
+        rc = spi_flash_write(pos, flash_write_puya_buf, bytesNow);
+        pos += bytesNow;
+    }
+    return rc;
+}
+#endif
+
 bool EspClass::flashWrite(uint32_t offset, uint32_t *data, size_t size) {
     ets_isr_mask(FLASH_INT_MASK);
-    int rc = spi_flash_write(offset, (uint32_t*) data, size);
+    int rc = 0;
+#if PUYA_SUPPORT
+    if (getFlashChipVendorId() == 0x85) { // 0x146085 PUYA
```

Comment:
as a courtesy to user who might be checking the same, please create a definition in Esp.h (as a macro `#define SPI_FLASH_VENDOR_PUYA 0x85` or a class member `static const uint8_t SPI_FLASH_VENDOR_PUYA = 0x85;`)

Reply:
I added all vendor IDs I could find in the Flashrom project.
I keep searching for these IDs myself, so I thought it would be best to included them all.